spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrTemplate.java

```diff
@@ -0,0 +1,250 @@
+
+/**
+ * Template providing convenient operations for interfacing with Google Cloud Vision's
+ * Document OCR feature, which allows you to run OCR algorithms on documents (PDF or TIFF format)
+ * stored on Google Cloud Storage.
+ *
+ * @author Daniel Zou
+ */
+public class DocumentOcrTemplate {
+
+	private static final Pattern OUTPUT_PAGE_PATTERN = Pattern.compile("output-\\d+-to-\\d+\\.json");
+
+	private final ImageAnnotatorClient imageAnnotatorClient;
+
+	private final Storage storage;
+
+	public DocumentOcrTemplate(
+			ImageAnnotatorClient imageAnnotatorClient,
+			Storage storage) {
+		this.imageAnnotatorClient = imageAnnotatorClient;
+		this.storage = storage;
+	}
+
+	/**
+	 * Runs OCR processing for a specified {@code document} and generates OCR output files
+	 * under the path specified by {@code outputFilePathPrefix}. One JSON output file is
+	 * produced for each page of the document.
+	 *
+	 * <p>
+	 * For example, if you specify an {@code outputFilePathPrefix} of
+	 * "gs://bucket_name/ocr_results/myDoc_", all the output files of OCR processing will be
+	 * saved under prefix, such as:
+	 *
+	 * <ul>
+	 * <li>gs://bucket_name/ocr_results/myDoc_output-1-to-1.json
+	 * <li>gs://bucket_name/ocr_results/myDoc_output-2-to-2.json
+	 * <li>gs://bucket_name/ocr_results/myDoc_output-3-to-3.json
+	 * </ul>
+	 *
+	 * <p>
+	 * Note: OCR processing operations may take several minutes to complete, so it may not be
+	 * advisable to block on the completion of the operation. One may use the returned
+	 * {@link ListenableFuture} to register callbacks or track the status of the operation.
+	 *
+	 * @param document The {@link GoogleStorageLocation} of the document to run OCR processing
+	 * @param outputFilePathPrefix The {@link GoogleStorageLocation} of a file, folder, or a
+	 *     bucket describing the path for which all output files shall be saved under
+	 *
+	 * @return A {@link ListenableFuture} allowing you to register callbacks or wait for the
+	 * completion of the operation.
+	 */
+	public ListenableFuture<DocumentOcrResultSet> runOcrForDocument(
+			GoogleStorageLocation document,
+			GoogleStorageLocation outputFilePathPrefix) {
+		if (!document.isFile()) {
+			throw new IllegalArgumentException(
+					"Provided document location is not a valid file location: " + document);
+		}
+
+		GcsSource gcsSource = GcsSource.newBuilder()
+				.setUri(document.uriString())
+				.build();
+
+		Blob documentBlob = this.storage.get(
+				BlobId.of(document.getBucketName(), document.getBlobName()));
+		if (documentBlob == null) {
+			throw new IllegalArgumentException(
+					"Provided document does not exist: " + document);
+		}
+
+		String contentType = documentBlob.getContentType();
+		InputConfig inputConfig = InputConfig.newBuilder()
+				.setMimeType(contentType)
+				.setGcsSource(gcsSource)
+				.build();
+
+		GcsDestination gcsDestination = GcsDestination.newBuilder()
+				.setUri(outputFilePathPrefix.uriString())
```

Comment:
The outputFilePathPrefix as very little restrictions (it can describe a file/folder/bucket) - in this case we can skip it because we can defer the error handling to the Vision API which will do validation on the URI.